Source/options.cpp

```diff
@@ -59,6 +59,40 @@ CSimpleIni &GetIni()
 	return ini;
 }
 
+static bool IniChanged = false;
+
+/**
+ * @brief Checks if a ini entry is changed by comparing value before and after
+ */
+struct IniChangedChecker {
+	IniChangedChecker(const char *sectionName, const char *keyName)
+	{
+		this->sectionName = sectionName;
+		this->keyName = keyName;
+		std::list<CSimpleIni::Entry> values;
+		if (!GetIni().GetAllValues(sectionName, keyName, values)) {
+			// No entry found in original ini => new entry => changed
+			IniChanged = true;
+		}
+		auto value = GetIni().GetValue(sectionName, keyName);
+		if (value != nullptr)
+			oldValue = value;
+	}
+	~IniChangedChecker()
+	{
+		auto value = GetIni().GetValue(sectionName, keyName);
+		std::string newValue;
+		if (value != nullptr)
+			newValue = value;
+		if (oldValue != newValue)
+			IniChanged = true;
+	}
+private:
+	std::string oldValue;
+	const char *sectionName;
+	const char *keyName;
+};

```

Comment:
The scope-based approach to this seems cool, but I'm a bit surprised by the apparent complexity of this class here.

Wouldn't it be simpler to have these checks at the model level? Say, after the values are loaded into the global settings model, have a way to tell which of the settings values have been updated, and then only save those?

Feels a bit weird to me having these checks at the actual "ini saving helpers" layer and not on the caller.

Reply:
Initially I thought the same. But the problem is that the "ini saving helpers" are not only responsible for saving the (user) changed values. They also generate the default ini (if not present). This is needed so the user can find what options he could change.

I thought about checking if the file exists. But if I did it this way, no new (default) entries would be added to the ini (for example `Hardware Cursor`).

That's why I came up with this solution.

After #1561 is resolved, we could revisit this, cause the reason for option discoverability would be gone.

Reply:
What if we distribute the default ini with the game package and stop "generating an empty one" at runtime?

I have always found it weird to generate the config in runtime. Coming from .NET, this is extremely unusual to me. Usually what happens is that the application has the configuration file as part of the project contents and this is deployed alongside the application.

Reply:
The ini that is created per default is not empty 😉 
And I think `diablo.ini` is more a settings file then classic `.config` file. Cause as you stated config files shouldn't normally be changed but distributed with the program.

Back to your question. 🙂 
The problem would be that we can't insert new entries for new options in existing ini files.

For example:

- User installed DevilutionX 1.21
- Ini is created with default values (copied ini file)
- User updates to DevilutionX 1.3
- Ini does exist, we don't copy cause that would override the users preferences => Missing `Hardware Cursor` entry in users `diablo.ini` file

Reply:
People love to strip everything except the .exe for some reason so if we can't boot without one we would probably get complaints.

Reply:
Fair enough.

Reply:
Also, yeah, didn't mean "empty" above, I know its saved with defaults initially.